File: goldenflow/config/learner.py
from __future__ import annotations

from pathlib import Path


from goldenflow.config.schema import GoldenFlowConfig, TransformSpec
from goldenflow.connectors.file import read_file
from goldenflow.engine.profiler_bridge import profile_dataframe
from goldenflow.engine.selector import select_transforms


def learn_config(path: Path) -> GoldenFlowConfig:
    """Profile a data file and generate a config from auto-detected transforms."""
    df = read_file(path)
    profile = profile_dataframe(df, file_path=str(path))

    transforms: list[TransformSpec] = []
    for col_profile in profile.columns:
        selected = select_transforms(col_profile)
        if selected:
            ops = [t.name for t in selected]
            transforms.append(TransformSpec(column=col_profile.name, ops=ops))

    return GoldenFlowConfig(
        source=str(path),
        transforms=transforms,
    )

File: packages/goldenflow-js/src/core/config/learner.ts
/**
 * Config learner — generate a config from data profiles.
 */

import type { GoldenFlowConfig, Row, TransformSpec } from "../types.js";
import { makeConfig } from "../types.js";
import { profileDataframe } from "../engine/profiler-bridge.js";
import { selectTransforms } from "../engine/selector.js";

export function learnConfig(rows: readonly Row[], source = ""): GoldenFlowConfig {
  const profile = profileDataframe(rows, source);
  const transforms: TransformSpec[] = [];

  for (const colProfile of profile.columns) {
    const selected = selectTransforms(colProfile);
    if (selected.length > 0) {
      transforms.push({
        column: colProfile.name,
        ops: selected.map((t) => t.name),
      });
    }
  }

  return makeConfig({
    source: source || null,
    transforms,
  });
}
